### PR TITLE
ASoC: Intel: Boards: tgl_max98373: add dai_trigger function

### PR DESCRIPTION
**Pull request by RDharageswari**
Speaker amplifier feedback is not modeled as being dependent on any
active output. Even when there is no playback happening, parts of the graph,
specifically the IV sense->speaker protection->output remains active and
this prevents the DSP from entering low-power states.

This patch suggest a machine driver level approach where the speaker
pins are enabled/disabled dynamically depending on stream start/stop
events. DPAM graph representations show the feedback loop is indeed
disabled and low-power states can be reached.

### CODE REVIEW

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 1]
```diff
@@ -57,8 +57,60 @@ static int max98373_hw_params(struct snd_pcm_substream *substream,
...
+				dev_err(rtd->dev, "failed to enable %s: %d\n",
+				pin_name, ret);
+				return ret;
+			}
+			snd_soc_dapm_sync(dapm);
+			break;
+		case SNDRV_PCM_TRIGGER_STOP:
+		case SNDRV_PCM_TRIGGER_SUSPEND:
+		case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
+			if (rtd->codec_dai->active != 1)
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_maxim_common.c]: @RDharageswari This field being removed and replaced with a new function. check out https://mailman.alsa-project.org/pipermail/alsa-devel/2020-May/167756.html.

Depending on when those patches are merged and this patch goes upstream, it might need updating

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 2]
```diff
@@ -57,8 +57,60 @@ static int max98373_hw_params(struct snd_pcm_substream *substream,
...
+			codec_dai->component->name_prefix);
+
+		switch (cmd) {
+		case SNDRV_PCM_TRIGGER_START:
+		case SNDRV_PCM_TRIGGER_RESUME:
+		case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
+			ret = snd_soc_dapm_enable_pin(dapm, pin_name);
+			if (ret) {
+				dev_err(rtd->dev, "failed to enable %s: %d\n",
+				pin_name, ret);
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_maxim_common.c]: we can remove this error and just break here and return ret at the end (this mean you will need to init ret as well). Same for the stop other triggers.

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 3]
```diff
@@ -57,8 +57,60 @@ static int max98373_hw_params(struct snd_pcm_substream *substream,
...
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dai *codec_dai;
+	int j, ret;
+
+	for_each_rtd_codec_dais(rtd, j, codec_dai) {
+		const char *name = codec_dai->component->name;
+		struct snd_soc_component *component = codec_dai->component;
+		struct snd_soc_dapm_context *dapm =
+				snd_soc_component_get_dapm(component);
+		char pin_name[20];
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_maxim_common.c]: why 20? can we use a macro here? or kasprintf()?

File: sound/soc/intel/boards/sof_rt5682.c [hunk 4]
```diff
@@ -349,7 +350,15 @@ static int sof_card_late_probe(struct snd_soc_card *card)
...
 	}
 
-	return hdac_hdmi_jack_port_init(component, &card->dapm);
+
+	err = hdac_hdmi_jack_port_init(component, &card->dapm);
+	if (err < 0)
+		return err;
+
+	snd_soc_dapm_disable_pin(dapm, "Left Spk");
+	snd_soc_dapm_disable_pin(dapm, "Right Spk");
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_rt5682.c]: Add a comment on why this is needed?

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/intel/boards/sof_maxim_common.c]: Also, codec_dai is gone from rtd, you need to use the macro asoc_rtd_to_codec(0). I suppose you're working with an older version of the kernel.

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 2 above)
Comment (RDharageswari) [reply sound/soc/intel/boards/sof_maxim_common.c]: Sure

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 3 above)
Comment (RDharageswari) [reply sound/soc/intel/boards/sof_maxim_common.c]: Sure

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 5]
```diff
@@ -57,8 +57,60 @@ static int max98373_hw_params(struct snd_pcm_substream *substream,
...
+
+		if (strcmp(name, MAX_98373_DEV0_NAME) &&
+			strcmp(name, MAX_98373_DEV1_NAME))
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_maxim_common.c]: Is this check really needed? I mean the trigger op is only set for the max_98373 BE DAI link, right?

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 6]
```diff
@@ -57,8 +57,47 @@ static int max98373_hw_params(struct snd_pcm_substream *substream,
...
+	int j, ret = 0;
+	char *pin_name;
+
+	for_each_rtd_codec_dais(rtd, j, codec_dai) {
+		struct snd_soc_component *component = codec_dai->component;
+		struct snd_soc_dapm_context *dapm =
+				snd_soc_component_get_dapm(component);
+
+		pin_name = devm_kasprintf(rtd->dev, GFP_KERNEL, "%s Spk",
+					codec_dai->component->name_prefix);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: this seems like a really bad idea to allocate a new string every time the trigger is invoked? The memory will only be released when the card is removed, so it's a good as just allocating memory without releasing it...

File: sound/soc/intel/boards/sof_rt5682.c [hunk 7]
```diff
@@ -349,6 +350,11 @@ static int sof_card_late_probe(struct snd_soc_card *card)
 		i++;
 	}
 
+	if (sof_rt5682_quirk & SOF_MAX98373_SPEAKER_AMP_PRESENT) {
+		snd_soc_dapm_disable_pin(dapm, "Left Spk");
+		snd_soc_dapm_disable_pin(dapm, "Right Spk");
+		return snd_soc_dapm_sync(dapm);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c]: don't return here, you will skip the hdmi initialization.

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 8]
```diff
@@ -57,8 +57,47 @@ static int max98373_hw_params(struct snd_pcm_substream *substream,
...
+		case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
+			ret = snd_soc_dapm_enable_pin(dapm, pin_name);
+			if (!ret)
+				snd_soc_dapm_sync(dapm);
+			break;
+		case SNDRV_PCM_TRIGGER_STOP:
+		case SNDRV_PCM_TRIGGER_SUSPEND:
+		case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
+			if (codec_dai->active != 1)
+				break;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c:91]: can you explain why the codec_dai would not be active when those cases occur? Is this test necessary?

Comment (RDharageswari) [reply sound/soc/intel/boards/sof_maxim_common.c:91]: During DSM, this is one of the sequence:
1)start playback
2)start a record stream(IV data)
3)record ends
4)playback ends.
pb & capture stream are on same SSP.
once the capture stream stops, the TRIGGER_STOP will be called for the capture and it will disable_pin if this condition check is not there.
This will cause no audio on playback stream even though it is running.
Hence the change

File: sound/soc/intel/boards/sof_rt5682.c (hunk 7 above)
Comment (RDharageswari) [reply sound/soc/intel/boards/sof_rt5682.c]: yeah..my bad, will fix this

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 6 above)
Comment (RDharageswari) [reply sound/soc/intel/boards/sof_maxim_common.c]: agree..will fix this

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 8 above)
Comment (kv2019i) [reply sound/soc/intel/boards/sof_maxim_common.c:91]: @RDharageswari So is "codec_dai->active==2" in this case? So this would essential ensure pin is disabled when the last codec dai user does trigger-stop, right?

Comment (RDharageswari) [reply sound/soc/intel/boards/sof_maxim_common.c:91]: Yes..

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 5 above)
Comment (RDharageswari) [reply sound/soc/intel/boards/sof_maxim_common.c]: removed

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 9]
```diff
@@ -57,8 +59,50 @@ static int max98373_hw_params(struct snd_pcm_substream *substream,
 	return 0;
 }
 
+static int max98373_trigger(struct snd_pcm_substream *substream, int cmd)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dai *codec_dai;
+	int j, ret = 0, val;
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_maxim_common.c]: int ret = 0; on a separate line please

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 10]
```diff
@@ -57,8 +59,50 @@ static int max98373_hw_params(struct snd_pcm_substream *substream,
...
+		case SNDRV_PCM_TRIGGER_RESUME:
+		case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
+			ret = snd_soc_dapm_enable_pin(dapm, pin_name);
+			if (!ret)
+				snd_soc_dapm_sync(dapm);
+			break;
+		case SNDRV_PCM_TRIGGER_STOP:
+		case SNDRV_PCM_TRIGGER_SUSPEND:
+		case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
+			val = snd_soc_dai_active(codec_dai);
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_maxim_common.c]: you dont seem to use val at all?

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 11]
```diff
@@ -27,11 +29,11 @@ static struct snd_soc_codec_conf max_98373_codec_conf[] = {
 };
 
 struct snd_soc_dai_link_component max_98373_components[] = {
-	{  /* For Left */
+	{  /* For Right */
 		.name = MAX_98373_DEV0_NAME,
 		.dai_name = MAX_98373_CODEC_DAI,
 	},
-	{  /* For Right */
+	{  /* For Left */
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_maxim_common.c:36]: the above 2 changes are unrelated. should be a separate patch or dropped if they're are not important

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 8 above)
Comment (ranj063) [reply sound/soc/intel/boards/sof_maxim_common.c:91]: can you please add the explanation as a comment before the check?

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 10 above)
Comment (RDharageswari) [reply sound/soc/intel/boards/sof_maxim_common.c]: done

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 11 above)
Comment (RDharageswari) [reply sound/soc/intel/boards/sof_maxim_common.c:36]: +1

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 9 above)
Comment (RDharageswari) [reply sound/soc/intel/boards/sof_maxim_common.c]: done

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 12]
```diff
@@ -9,6 +9,8 @@
 #include <uapi/sound/asound.h>
 #include "sof_maxim_common.h"
 
+#define MAX_98373_PIN_NAME 16
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_maxim_common.c:12]: please remove change id from commit message and fix typo "suggest" -> "suggests"

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 8 above)
Comment (RDharageswari) [reply sound/soc/intel/boards/sof_maxim_common.c:91]: Sure

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 12 above)
Comment (RDharageswari) [reply sound/soc/intel/boards/sof_maxim_common.c:12]: Sure

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 13]
```diff
@@ -57,8 +59,51 @@ static int max98373_hw_params(struct snd_pcm_substream *substream,
...
+	int ret = 0;
+
+	for_each_rtd_codec_dais(rtd, j, codec_dai) {
+		struct snd_soc_component *component = codec_dai->component;
+		struct snd_soc_dapm_context *dapm =
+				snd_soc_component_get_dapm(component);
+		char pin_name[MAX_98373_PIN_NAME];
+
+		snprintf(pin_name, ARRAY_SIZE(pin_name), "%s Spk",
+			 codec_dai->component->name_prefix);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c:76]: does this work if you have no name prefix? the pin name would have an extra space at the beginning, Or is there always a component prefix?

Comment (RDharageswari) [reply sound/soc/intel/boards/sof_maxim_common.c:76]: there is a component name define here in the same file, so there will always be a component prefix.
